function-sidecar/pkg/dispatcher/grpc/grpc.go

```diff
@@ -52,59 +58,157 @@ func (this *grpcDispatcher) Closed() <-chan struct{} {
 	return this.closed
 }
 
-func (this *grpcDispatcher) handleIncoming() {
+// A CorrelationStrategy groups together messages that belong to the same "spatial" stream (the "temporal" aspect
+// of grouping is taken care of by a WindowStrategy). Several invocations for different spatial groups may happen concurrently.
+type CorrelationStrategy func(message.Message) interface{}
+
+// A WindowingStrategyFactory is a function for creating a new instance of a (most certainly stateful) WindowingStrategy,
+// given a new, first message for a stream.
+type WindowingStrategyFactory func(message.Message) WindowingStrategy
+
+//
+type WindowingStrategy interface {
+	// ShouldClose notifies this strategy that a new message has just been sent to the function, resulting in the optional
+	// error passed as 2nd argument. This is a synchronous way of deciding whether or not a window should be closed.
+	// ShouldClose should return true if this strategy decides that the current stream should end.
+	ShouldClose(in message.Message, err error) bool
+
+	// AsyncClose should return a non-nil channel that this strategy will close if it decides that the current window
+	// should end asynchronously
+	AsyncClose() <-chan struct{}
```

Comment:
Wouldn't it be cleaner to return the channel from the factory since the channel and the `WindowingStrategy` are essentially peers of each other? A benefit of this approach would be that you could then build a `select` from a fixed number of cases which will make the loop below easier to read.

Reply:
Could do.

The select will not be out of fixed number of cases though, because streams come and go at runtime, so will always need to deal with that.

But optional channel as second return value is indeed more sensible

Reply:
Keeping the channel as the way it is now because:
- it needs to be looked at more than once, not just at construction time
- feels weird to have an XFactory return an X and something else, which really belongs to the logic of X..